Indexeddb: transaction lifecycle

Taym95 opened the pull request:
Indexeddb: transaction lifecycle

Testing: more Indexeddb test should pass
part of #40983

gterzian commented on components/storage/indexeddb/mod.rs:
```diff
@@ -194,6 +240,134 @@ impl<E: KvsEngine> IndexedDBEnvironment<E> {
+                }
+                // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+                // the implementation must queue a database task to start the transaction asynchronously.
+                let _ = sender.send(IndexedDBThreadMsg::StartQueuedTransaction {
```
I don't understand why the spec says to queue a task, since this seems to be a message to the backend?

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -49,11 +50,25 @@ pub struct IDBTransaction {
+    // are run at most once for each transaction.”
+    // https://w3c.github.io/IndexedDB/#cleanup-indexed-database-transactions
+    #[no_trace]
+    cleanup_event_loop: DomRefCell<Option<ScriptEventLoopId>>,
```
So I don't think we need the `ScriptEventLoopId`, and also we shouldn't store this on the transaction. Instead we should store transactions on the global, and then clean-up transactions for the global at the microtask checkoint: see step 5 of https://html.spec.whatwg.org/#perform-a-microtask-checkpoint.

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
+            .map(|request| Trusted::new(&**request))
+            .collect();
+        let global = self.global();
+        global.task_manager().database_access_task_source().queue(
```
Why queue a task since this is called already from a task? I'm guessing you need to first abort on the backend, and then send a message back to script to queue a task to fire the events.

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
+        self.pending_request_count.set(pending - 1);
+    }
+
+    fn maybe_finish_successfully(&self) {
```
Again I think you first need to wait on the backend for all requests to have completed, and then communicate back to script to continue the finish logic here.

gterzian commented on components/script/dom/indexeddb/idbrequest.rs:
```diff
@@ -46,6 +46,17 @@ struct RequestListener {
+
+impl Drop for TxnActiveGuard<'_> {
+    fn drop(&mut self) {
+        self.0.set_active_flag(false);
```
So I don't think you need this, essentially you just need to set it to non-active after firing the event.

gterzian commented on components/shared/storage/indexeddb.rs:
```diff
@@ -488,9 +489,25 @@ pub enum IndexedDBThreadMsg {
         IndexedDBTxnMode,
         AsyncOperation,
     ),
+    StartQueuedTransaction {
```
This also should be determined by the backend; no need for a message.

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
+        self.upgrade_old_version.set(Some(old_version));
+    }
+
+    pub(crate) fn initiate_abort(&self, error: Error, can_gc: CanGc) {
```
https://w3c.github.io/IndexedDB/#abort-a-transaction

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
+        // https://w3c.github.io/IndexedDB/#abort-a-transaction
+        // When a transaction is aborted the implementation must undo (roll back)
+        // any changes that were made to the database during that transaction.”
+        self.queue_abort_pending_requests();
```
So the spec is weird but I think you should only send the message above, do the abort on the backend, and then continue in script for this and the below.

gterzian commented on components/script/dom/indexeddb/idbrequest.rs:
```diff
@@ -239,11 +256,18 @@ impl RequestListener {
-            transaction.set_active_flag(false);
+
+            // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+            // A transaction is in this state after control returns to the event loop after its creation,
```
So this points to the inactive state, not finished as below.

gterzian reviewed (commented):
So I think it's a good start but I think some structural pieces are missing. 

At a high-level(mostly for my own understanding):

When a transaction is created, it is set to active. Then at the next microtask it is set to inactive. However, for each even that fires after that for a request, it is set back to active. After the event has been fired and control is back to Rust, it should be set back to inactive. 

The concept of "starting" a transaction I think is mostly a backend thing, even though the spec queues a task for it. Basically, while a transaction is active, requests can be made against it, but those requests should only start when the transaction can be started. 

I added some comments here and there about structural improvements. Mostly about doing something via a round-trip with the backend as oppposed to directly in script.

Taym95 committed:
IndexedDB Transaction lifecycle

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -159,19 +180,133 @@ impl IDBTransaction {
+            },
+        )
+        .expect("Could not create callback");
+        let start_operation = SyncOperation::Commit(
```
commit operation

gterzian reviewed (commented):
First pass at the overall structure. 

cc @janvarga can you pls take a look also?

Taym95 committed:
address review comments

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Fix clippy

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

gterzian reviewed (commented):
couple of more notes; any idea on the timeouts?

Taym95 commented:
> couple of more notes; any idea on the timeouts?

I am debugging them

Taym95 committed:
Add verbose debug logs and fix several timeout

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Remove println and update test expectations

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Fix more timeout test

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Align with spec and pass more tests

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Fix idbtransaction-objectStore-finished test

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
More tests passing

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Fix more transaction-lifecycle tests

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Make more transaction-lifecycle tests pass

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
avoid spawning a thread each time process_transaction

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

gterzian commented on components/script/dom/indexeddb/idbdatabase.rs:
```diff
@@ -105,26 +108,67 @@ impl IDBDatabase {
+        let queue = self.transaction_queue.borrow();
+        let at_head = queue
+            .front()
+            .is_some_and(|front| std::ptr::eq::<IDBTransaction>(&**front, transaction));
```
The pointer stuff is needed?

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -160,23 +185,183 @@ impl IDBTransaction {
+        // any of the changes to the database, and the steps to abort a transaction will be followed.
+        let send_result = self
+            .get_idb_thread()
+            .send(IndexedDBThreadMsg::Sync(commit_operation));
```
Can we not run the equivalent logic on the backend in response to receiving the request handled message(which could be re-used for all inactive cases)?

gterzian reviewed (commented):
LGTM generally speaking, although I expected more logic to run, and lifecycle bookkeeping to be, on the backend, but this sounds more like something to investigate in a follow-up than now. 

Couple of comments to address.

Taym95 commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -51,19 +55,31 @@ pub struct IDBTransaction {
     /// The id of the associated open request, if any.
     #[no_trace]
-    open_request_id: Option<Uuid>,
+    _open_request_id: Option<Uuid>,
```
this was part of my debug logs, I will remove it

arihant2math commented:
I'm a bit worried about the intermittent crashes, as they tend to be around 1/15-1/20 in occurrence and occur with a large subset of tests. I don't know if it's worth increasing intermittency by that much.

arihant2math commented:
#42444 turns them into timeouts, I'll try and look into the root cause, but atleast a timeout is marginally better.

arihant2math commented:
I'm pretty sure these changes are fine, but they uncover some underlying problem.

Taym95 commented on components/script/dom/indexeddb/idbdatabase.rs:
```diff
@@ -44,6 +45,7 @@ pub struct IDBDatabase {
     object_store_names: DomRefCell<Vec<DOMString>>,
     /// <https://w3c.github.io/IndexedDB/#database-upgrade-transaction>
     upgrade_transaction: MutNullableDom<IDBTransaction>,
+    transaction_queue: DomRefCell<VecDeque<Dom<IDBTransaction>>>,
```
I will reanme it to `commit_order_queue` and add doc comemnt.

Taym95 committed:
Enforce spec transaction scheduling in backend

Implement §2.7.2 overlap/created-before scheduling in the backend by tracking
txn metadata (mode/scope/created order). Register txn metadata from script at
creation time, remove script-side txn queue, and add TxnMaybeCommit + explicit
TransactionFinished to keep commit checks and liveness correct.

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
fix upgrade-transaction-lifecycle-committed test

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Merge branch 'servo-main' into IndexedDB-transaction-lifecycle

Taym95 committed:
Update MANIFEST.json

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

gterzian commented on components/script/dom/indexeddb/idbfactory.rs:
```diff
@@ -58,10 +63,101 @@ impl IDBFactory {
+                .collect()
+        };
+        for request in requests {
+            request.clear_transaction_if_matches(transaction);
```
So here I think there should only ever be one, so maybe add a debug assert that `requests` len is 1.

gterzian commented on components/storage/indexeddb/mod.rs:
```diff
@@ -796,7 +1232,7 @@ impl IndexedDBManager {
                 version: new_version,
                 old_version,
-                transaction: transaction_id,
+                transaction: 0,
```
remove?

gterzian commented on components/storage/indexeddb/mod.rs:
```diff
@@ -767,9 +1206,6 @@ impl IndexedDBManager {
 
         // Step 2: Let transaction be a new upgrade transaction with connection used as connection.
-        let transaction_id = self.serial_number_counter;
-        self.serial_number_counter += 1;
```
Here would be good to add a note on how we link the transaction with the open request, my guess is: since there can be only one open requests at a time, and one transction associated with it, it works without having to pass around the `transaction_id`?

gterzian commented on components/script/dom/indexeddb/idbtransaction.rs:
```diff
@@ -85,98 +95,270 @@ impl IDBTransaction {
+        }
+        // Commit must respect backend scheduling for overlapping scopes.
+        // https://w3c.github.io/IndexedDB/#transaction-scheduling
+        if !self.can_start_in_backend() {
```
I would like to see this blocking call dissapear: for abort it seems you can always proceed; for commit can you not run this check as part of handing the `SyncOperation::Commit`?

Would be good to do in a follow-up.

gterzian reviewed (approved):
OK LGTM with some comments and so follow-ups to note in the meta issue. 

Great work!

Taym95 committed:
move commit scheduling to backend and clean up upgrade txn wiring

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Fix clippy

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
Use stepwise type of docs

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

janvarga commented on components/script/dom/indexeddb/idbdatabase.rs:
```diff
@@ -105,26 +108,67 @@ impl IDBDatabase {
+        // IndexedDB §5.8 notes that aborting an upgrade reverts this returned value.
+        // https://w3c.github.io/IndexedDB/#dom-idbdatabase-version
+        // https://w3c.github.io/IndexedDB/#abort-an-upgrade-transaction
+        self.version.get()
```
Right, IDBDatabase::version can't dynamically change AFAIK (once it's set).

Taym95 committed:
Merge branch 'servo-main' into IndexedDB-transaction-lifecycle

janvarga commented on components/script/dom/indexeddb/idbdatabase.rs:
```diff
@@ -185,7 +185,23 @@ impl IDBDatabaseMethods<crate::DomTypeHolder> for IDBDatabase {
+            self.global()
+                .get_indexeddb()
+                .register_indexeddb_transaction(&transaction);
+            transaction.set_registered_in_global();
```
It looks like register_indexeddb_transaction/unregister_indexeddb_transaction could also set the flag (registered_in_global) and be a bit less error prone?

Taym95 replied:
indeed this is better.

janvarga reviewed (commented):
Submitting some comments, I should be done with the review in couple of minutes.

janvarga reviewed (approved):
Nice work!

This is a fairly large set of changes, so I didn’t manage to verify every single detail, but from a high-level perspective this looks good to me (modulo a few inline comments and questions).

Taym95 commented on components/script/dom/indexeddb/idbfactory.rs:
```diff
@@ -58,10 +63,105 @@ impl IDBFactory {
+        self.indexeddb_transactions
+            .borrow_mut()
+            .retain(|txn| !txn.is_finished());
+        true
```
So the snapshot gives me a clean separation:

- Borrow list briefly ->> clone roots into `Vec<DomRoot<_>>`

- Drop borrow

- Work on the snapshot safely (calls may re enter)

Taym95 commented on components/script/dom/indexeddb/idbrequest.rs:
```diff
@@ -396,11 +473,12 @@ impl IDBRequest {
         let task_source = global
             .task_manager()
-            .database_access_task_source()
+            .dom_manipulation_task_source()
```
this should be database_access_task_source, I will undo it

Taym95 committed:
IndexedDB: address review notes on txn tracking, task sources, and scheduling

- Keep a defensive prune of finished transactions before snapshotting in cleanup, since unregister can happen in queued tasks.
- Rework script-side transaction tracking to be keyed by db_name to avoid linear scans when looking up a txn to maybe-commit.
- Use dom_manipulation_task_source where required to preserve ordering/expectations; document why database_access_task_source caused test failures.
- Rename handled_next_unhandled_id to handled_next_unhandled_request_id for clarity (aligns with backend terminology).
- Treat backend commit failure mapping as temporary; avoid implying QuotaExceeded unless the backend reports it.
- Improve backend scheduling performance by avoiding O(n) membership checks on queues (keep VecDeque for FIFO order and add a set for O(1) contains).

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Taym95 committed:
Fix clippy

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 committed:
If the database entry has already been removed, treat commit as a
 no-op success so script side completion does not spuriously abort

Signed-off-by: Taym Haddadi <haddadi.taym@gmail.com>

Taym95 commented on components/script/dom/indexeddb/idbdatabase.rs:
```diff
@@ -185,7 +185,23 @@ impl IDBDatabaseMethods<crate::DomTypeHolder> for IDBDatabase {
             },
-        })
+        };
+        if mode != IDBTransactionMode::Versionchange {
```
This made another test pass.
